contrastive_vi/module/total_contrastive_vi.py

```diff
@@ -0,0 +1,738 @@
+"""PyTorch module for Contrastive VI for single cell expression data."""
+
+from typing import Dict, Optional, Tuple, Union, Literal, Iterable
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from scvi import REGISTRY_KEYS
+from scvi.distributions import NegativeBinomial, NegativeBinomialMixture
+from scvi.module.base import BaseModuleClass, LossRecorder, auto_move_data
+from scvi.nn import DecoderTOTALVI, EncoderTOTALVI, one_hot
+from torch.distributions import Normal
+from torch.distributions import kl_divergence as kl
+
+torch.backends.cudnn.benchmark = True
+
+
+class TotalContrastiveVIModule(BaseModuleClass):
+    """
+    PyTorch module for total-contrastiveVI (contrastive analysis for CITE-seq).
+
+    Args:
+    ----
+        n_input_genes: Number of input genes.
+        n_input_proteins: Number of input proteins.
+        n_batch: Number of batches. If 0, no batch effect correction is performed.
+        n_hidden: Number of nodes per hidden layer.
+        n_background_latent: Dimensionality of the background latent space.
+        n_salient_latent: Dimensionality of the salient latent space.
+        n_layers: Number of hidden layers used for encoder and decoder NNs.
+        dropout_rate: Dropout rate for neural networks.
+        protein_batch_mask: Dictionary where each key is a batch code, and value is for
+            each protein, whether it was observed or not.
+        use_observed_lib_size: Use observed library size for RNA as scaling factor in
+            mean of conditional distribution.
+        library_log_means: 1 x n_batch array of means of the log library sizes.
+            Parameterize prior on library size if not using observed library size.
+        library_log_vars: 1 x n_batch array of variances of the log library sizes.
+            Parameterize prior on library size if not using observed library size.
+    """
+
+    def __init__(
```

Comment:
Similar to the `TotalContrastiveVIModel` class, I updated the `__init__` function here to match changes in the newest version of the totalVI implementation in `scvi-tools`. As a result, much of the code here up until line 166 is new (though almost everything is copy-pasted from the scvi-tools repository).